crates/rome_js_analyze/src/analyzers/complexity/no_confusing_void_type.rs

```diff
@@ -0,0 +1,281 @@
+
+declare_rule! {
+    ///
+    /// Disallow void type outside of generic or return types.
+    ///
+    /// void in TypeScript refers to a function return that is meant to be ignored. Attempting to use a void type outside of a return type or generic type argument is often a sign of programmer error. void can also be misleading for other developers even if used correctly.
+    ///
+    /// > The void type means cannot be mixed with any other types, other than never, which accepts all types. If you think you need this then you probably want the undefined type instead.
+    ///
+    /// ## Examples
+    /// ### Invalid
+    ///
+    /// ```ts,expect_diagnostic
+    /// type PossibleValues = number | void;
+    /// type MorePossibleValues = string | ((number & any) | (string | void));
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// function logSomething(thing: void) {}
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// interface Interface {
+    ///     prop: void;
+    /// }
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// let foo: void;
+    /// let bar = 1 as unknown as void;
+    /// let baz = 1 as unknown as void | string;
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```ts
+    /// function foo(): void {};
+    /// ```
+    ///
+    /// ## Options
+    ///
+    /// ### allowInGenericTypeArguments
+    ///
+    /// This option lets you control if void can be used as a valid value for generic type parameters.
+    ///
+    /// This option is `true` by default.
+    ///
+    /// The following patterns are considered warnings with `{ allowInGenericTypeArguments: false }`:
+    ///
+    /// ```ts
+    /// logAndReturn<void>(undefined);
+    ///
+    /// let voidPromise: Promise<void> = new Promise<void>(() => {});
+    /// let voidMap: Map<string, void> = new Map<string, void>();
+    /// ```
+    ///
+    /// ### allowAsThisParameter
+    ///
+    ///
+    ///
+    ///
+    /// This option allows specifying a this parameter of a function to be void when set to true. This pattern can be useful to explicitly label function types that do not use a this argument. See the TypeScript docs for more information.
+    /// This option is `false` by default.
+    ///
+    /// The following patterns are considered warnings with `{ allowAsThisParameter: false }` but valid with `{ allowAsThisParameter: true }`:
+    ///
+    /// ```ts
+    /// function doThing(this: void) {}
+    /// class Example {
+    ///     static helper(this: void) {}
+    ///     callback(this: void) {}
+    /// }
+    /// ```
+    ///
+    pub(crate) NoConfusingVoidType {
+        version: "1.0.0",
+        name: "noConfusingVoidType",
+        recommended: false,
+    }
+}
+
+/// rule options for `noConfusingVoidType`
+#[derive(Debug, Clone, Deserialize, Serialize, Bpaf)]
+#[cfg_attr(feature = "schema", derive(JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct NoConfusingVoidTypeOptions {
```

Comment:
Biome tries minimizing options.
I think we can remove these two options and assume that they are set to `true`.